src/core/qgsmaprenderercustompainterjob.cpp

```diff
@@ -308,11 +311,47 @@ void QgsMapRendererCustomPainterJob::doRender()
       mLabelJob.participatingLayers = _qgis_listRawToQPointer( mLabelingEngineV2->participatingLayers() );
     }
   }
-  if ( mLabelJob.img && mLabelJob.complete )
+
+  if ( ! hasSecondPass )
   {
+    if ( mLabelJob.img && mLabelJob.complete )
+    {
+      mPainter->setCompositionMode( QPainter::CompositionMode_SourceOver );
+      mPainter->setOpacity( 1.0 );
+      mPainter->drawImage( 0, 0, *mLabelJob.img );
+    }
+  }
+  else
+  {
+    for ( LayerRenderJob &job : mSecondPassLayerJobs )
+    {
+      if ( job.context.renderingStopped() )
+        break;
+
+      if ( !job.cached )
+      {
+        QTime layerTime;
+        layerTime.start();
+
+        if ( job.img )
+        {
+          job.img->fill( 0 );
+          job.imageInitialized = true;
+        }
+
+        job.renderer->render();
+
+        job.renderingTime += layerTime.elapsed();
+      }
+    }
+
+    composeSecondPass( mSecondPassLayerJobs, mLabelJob );
```

Comment:
I'm trying to digest the consequences of this -- does this mean that when masks are used, labels are always composed onto a raster image? (ie. we lose the ability to get vector or text based label outputs?)

Reply:
I think so yes. I would have also preferred a vector based solution (with some QPainterPath that define clipping paths), but it seemed even more complex (and probably slower) than the raster based approach.

Reply:
I'm totally ok with the mask itself being a raster -- I think that's almost unavoidable. But is the actual label text itself rasterised? I can't tell from the code

Reply:
The rendering is done more or less following the parallel rendering process: each layer (including the "label" pseudo-layer) is drawn in its own QImage. Then some images are modified after a second pass, and finally all images are "stacked" to compose the final image. The sequential renderer does something equivalent when masks are involved.

So I guess the answer is yes: labels are rasterised.

That might be avoidable for the labeling part (since labels are always on top), but would require to render them twice : one time for the mask and another time for the final rendering.